### PR TITLE
Strict memory management checks for non-classes

### PR DESCRIPTION
**Pull request by DanilaFe**
Realized I forgot to open this during the day.

Closes https://github.com/chapel-lang/chapel/issues/14121. Closes https://github.com/chapel-lang/chapel/issues/25549.

This PR ensures that errors are thrown when non-class types are used with `shared` and `owned`. The previous error was limited to invocations of the initializers for `_owned` / `_shared`, which does not happen in certain cases, including when using type aliases and specifying formal types. This PR adds special-case logic for management types to the type instantiation code directly, which guarantees that it's run as soon as `owned(T)` is attempted. As a result, errors are now caught earlier and in more cases.

Reviewed by @jabraham17 -- thanks!

## Testing
- [x] paratest + new tests

### CODE REVIEW

File: compiler/AST/AggregateType.cpp [hunk 1]
```diff
@@ -986,6 +986,22 @@ AggregateType* AggregateType::generateType(CallExpr* call,
...
+  // Don't even allow creating an instantiation of something like
+  // 'owned' with a non-class.
+  bool isManagedPtrType = symbol->hasFlag(FLAG_MANAGED_POINTER);
+  if (isManagedPtrType && genericFields.size() > 0) {
+    // Assume first generic field is the underlying type.
+    auto field = genericFields[0];
+
+    if (auto sub = toTypeSymbol(map.get(field))) {
+      if (!isClassLike(sub->typeInfo())) {
+        USR_FATAL(call, "cannot use type non-class type '%s' with memory management specifiers",
```
Comment (jabraham17) [inline compiler/AST/AggregateType.cpp:998]: the double usage of "type" doesn't parse right.

I suggest just removing the first "type"

**Review by jabraham17 — APPROVED** [review]
Looks good to me.

I believe this PR removes the need for the "___ only works with classes" error messages in module code in OwnedObject and SharedObject

```chapel
    if !isClass(chpl_t) then
      compilerError("owned only works with classes");
```
Maybe its worth at tagging those call sites with `// TODO: remove me ` or the like. I am hesitant to remove them now, unless you can validate that they truly aren't needed anymore.

File: compiler/AST/AggregateType.cpp (hunk 1 above)
Comment (DanilaFe) [reply compiler/AST/AggregateType.cpp]: Yeah I just noticed that one.

**Comment by DanilaFe** [comment]
> Maybe its worth at tagging those call sites with // TODO: remove me or the like. I am hesitant to remove them now, unless you can validate that they truly aren't needed anymore.

I had the same though, and just like you I was hesitant to remove it. I'll add a comment.

File: modules/internal/SharedObject.chpl [hunk 2]
```diff
@@ -135,6 +135,9 @@ module SharedObject {
   pragma "leaves this nil"
   @chpldoc.nodoc // hide init/record impl details
   proc _shared.init(type chpl_t) {
+    // TODO: today (06/15/2024), the compiler has a special check for a non-class type
+    // being used to instnatiate _shared, so this check is likely redundant and
+    // should be removed.
     if !isClass(chpl_t) then
```
Comment (jabraham17) [inline modules/internal/SharedObject.chpl:141]: There are 2 other `shared.init` that need the same comment

**Review by jabraham17 — COMMENTED** [review]
Since you modified `OwnedObject` and `SharedObject`, I recommend rerunning a paratest to make sure line numbers still match.

I think `test/functions/fcf/ParseCastPrecedence.bad`, `test/studies/adventOfCode/2021/bradc/futures/day18-bug2.bad`, and `test/classes/delete-free/owned/errors/owned-int-type.bad` need to be updated.

**Comment by DanilaFe** [comment]
Looks like all the tests you mentioned have a prediff; a paratest is clean.

**Pull request merged**